src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -4987,8 +4967,13 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
         }
       }
 
-      def typedTypeSelectionQualifier(tree: Tree, pt: Type = AnyRefTpe) =
-        context.withImplicitsDisabled { typed(tree, MonoQualifierModes | mode.onlyTypePat, pt) }
+      def typedTypeSelectionQualifier(tree: Tree, pt: Type = AnyRefTpe) = {
+        def typedQual = context.withImplicitsDisabled { typed(tree, MonoQualifierModes | mode.onlyTypePat, pt) }
+        tree match {
+          case id: Ident if context.unit.isJava => context.withinJavaTypeSelection { typedQual }
```

Comment:
Talking this through with @retronym, it occurred to us that this `withinJavaTypeSelection` context bit would also be useful in @hrhino's quest to improve error messages in #6042. With that insight, should it just be `withinTypeSelection` (and should we have one for `withinPattern`?)

Also, why is this is only done when we reach an `Ident`? What are the varying combos of `A`, `A.B` and `A.B.C`.

Reply:
> why is this is only done when we reach an Ident

For selections there's an existing fix / workaround, see my comment here https://github.com/scala/scala/pull/6053#issuecomment-328479876.

Reply:
Would it be feasible to combine the two fixes?

Reply:
Yes, I think so.

Reply:
@adriaanm i was having a look at rebasing this PR and addressing your comment. Do you want me to combine the two fixes on this PR or how should we operate? /cc @hrhino

Reply:
combining sounds good to me

Reply:
Alright, I'll cherry-pick @hrhino's changes into this PR :+1: